File: app_java/src/main/java/cn/xhuww/recyclerview/itemanimator/LeftInRightOutAnimator.java
package cn.xhuww.recyclerview.itemanimator;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewPropertyAnimator;

public class LeftInRightOutAnimator extends BaseItemAnimator {

    @Override
    void prepareAnimateAddImpl(RecyclerView.ViewHolder viewHolder) {
        super.prepareAnimateAddImpl(viewHolder);
        View view = viewHolder.itemView;
        view.setTranslationX(-view.getWidth());
    }

    @Override
    void animateAddImpl(RecyclerView.ViewHolder viewHolder, ViewPropertyAnimator animation) {
        animation.setDuration(getAddDuration()).alpha(1F).translationX(0F);
    }

    @Override
    void animateRemoveImpl(RecyclerView.ViewHolder viewHolder, ViewPropertyAnimator animation) {
        animation.setDuration(getRemoveDuration()).translationX(-viewHolder.itemView.getWidth());
    }
}

File: app_kotlin/src/main/java/cn/xhuww/recyclerview/itemanimator/LeftInRightOutAnimator.kt
package cn.xhuww.recyclerview.itemanimator

import android.support.v7.widget.RecyclerView
import android.view.ViewPropertyAnimator

class LeftInRightOutAnimator : CommonItemAnimator() {

    override fun prepareAnimateAddImpl(viewHolder: RecyclerView.ViewHolder) {
        val view = viewHolder.itemView
        view.translationX = -view.width.toFloat()
    }

    override fun animateAddImpl(
        viewHolder: RecyclerView.ViewHolder, animation: ViewPropertyAnimator
    ) {
        animation.alpha(1F).translationX(0F)
    }

    override fun animateRemoveImpl(
        viewHolder: RecyclerView.ViewHolder, animation: ViewPropertyAnimator
    ) {
        animation.translationX(-viewHolder.itemView.width.toFloat())
    }
}
